src/ImageSharp.Web.Providers.Azure/Providers/BlobStorageImageProvider.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Extensions;
+using Microsoft.Extensions.Options;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Blob;
+using SixLabors.ImageSharp.Web.Helpers;
+using SixLabors.ImageSharp.Web.Middleware;
+using SixLabors.ImageSharp.Web.Resolvers;
+
+namespace SixLabors.ImageSharp.Web.Providers
+{
+    /// <summary>
+    /// Returns images stored in Azure Blob Storage.
+    /// </summary>
+    public class BlobStorageImageProvider : IImageProvider
+    {
+        /// <summary>
+        /// The cloud storage account.
+        /// </summary>
+        private readonly CloudStorageAccount storageAccount;
+
+        /// <summary>
+        /// The container in the blob service.
+        /// </summary>
+        private readonly CloudBlobContainer container;
+
+        /// <summary>
+        /// The middleware configuration options.
+        /// </summary>
+        private readonly ImageSharpMiddlewareOptions options;
+
+        /// <summary>
+        /// The blob storage options.
+        /// </summary>
+        private readonly BlobStorageImageProviderOptions storageOptions;
+
+        /// <summary>
+        /// Contains various helper methods based on the current configuration.
+        /// </summary>
+        private readonly FormatHelper formatHelper;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BlobStorageImageProvider"/> class.
+        /// </summary>
+        /// <param name="options">The middleware configuration options.</param>
+        /// <param name="storageOptions">The blob storage options.</param>
+        public BlobStorageImageProvider(IOptions<ImageSharpMiddlewareOptions> options, IOptions<BlobStorageImageProviderOptions> storageOptions)
+        {
+            Guard.NotNull(options, nameof(options));
+            Guard.NotNull(storageOptions, nameof(storageOptions));
+
+            this.options = options.Value;
+            this.storageOptions = storageOptions.Value;
+            this.formatHelper = new FormatHelper(this.options.Configuration);
+            this.storageAccount = CloudStorageAccount.Parse(this.storageOptions.ConnectionString);
+
+            CloudBlobClient client = this.storageAccount.CreateCloudBlobClient();
```

Comment:
Have you tested (or otherwise confirmed) that the CloudBlobClient and CloudBlobContainer classes are thread safe?  The Azure Storage documentation is extremely vague on this topic (see [here](https://github.com/Azure/azure-storage-net/issues/732)) and it wasn't clear to me whether they can be created once here and then re-used in parallel by multiple different requests, or if they should be created per-request.

Reply:
I've used a static `CloudBlobClient` before in ImageProcessor. Creating on per request will likely exhaust sockets. I'm confident that since we are not modifying the actual client we are safe.